Construct a TimeHandle using blockfrost

v0d1ch opened the pull request:
- As a step towards blockfrost integration we should be able to construct a `TimeHandle` in order to do slot/time conversions.
- Depends on work in https://github.com/cardano-scaling/hydra/pull/1934

---

<!-- Consider each and tick it off one way or the other -->
* [x] CHANGELOG updated or not needed
* [x] Documentation updated or not needed
* [x] Haddocks updated or not needed
* [x] No new TODOs introduced or explained herafter

v0d1ch committed:
Put Wallet types into separate module

Also introduce Hydra.Chain.Blockfrost.Wallet

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Put Wallet types into separate module

Also introduce Hydra.Chain.Blockfrost.Wallet

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Put Wallet types into separate module

Also introduce Hydra.Chain.Blockfrost.Wallet

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Put Wallet types into separate module

Also introduce Hydra.Chain.Blockfrost.Wallet

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Put Wallet types into separate module

Also introduce Hydra.Chain.Blockfrost.Wallet

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Introduce queryWalletInfo and new module Hydra.Chain.Blockfrost

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Introduce queryWalletInfo and new module Hydra.Chain.Blockfrost

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Introduce queryEpochInfo and getSeedInput BF variant

Add haddocks for blockfrost client functions
Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Wallet update function and query PParams

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Cleanup

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Simplify blockfrost setup

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Simplify blockfrost setup

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Simplify blockfrost setup

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Simplify blockfrost setup

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Simplify blockfrost setup

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Simplify blockfrost setup

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Update changelog

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Update changelog

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Construct EraHistory correctly

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Construct a TimeHandle using blockfrost

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Construct a TimeHandle using blockfrost

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Construct a TimeHandle using blockfrost

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Construct a TimeHandle using blockfrost

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

ch1bo commented on hydra-node/src/Hydra/Chain/Blockfrost/TimeHandle.hs:
```diff
@@ -0,0 +1,34 @@
+import Hydra.Chain.CardanoClient (
+  QueryPoint (QueryTip),
+ )
+import Hydra.Chain.Direct.TimeHandle (TimeHandle, mkTimeHandle)
```
I'd suggest to move `TimeHandle` (and also `Wallet`) into `Hydra.Chain` to indicate its used by both, `Direct` and `Blockfrost` chain backends.

v0d1ch replied:
It is annoying to move the `TimeHandle` because of the arbitrary instance that relies on `newTinyWallet`

ch1bo replied:
`Arbitrary TimeHandle` does not depend on `newTinyWallet`??

This is the instance on `master`:
```haskell
instance Arbitrary TimeHandle where
  arbitrary = do
    TimeHandleParams{systemStart, eraHistory, currentSlot} <- genTimeParams
    pure $ mkTimeHandle currentSlot systemStart eraHistory

```

v0d1ch committed:
Remove changelog entry as it is not user facing

Some fixes after rebase

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Remove squash in favor of fold

fold already has a Monoid and Foldable constraints so it is convenient.

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Draft common types between BF/Cardano

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

ch1bo committed:
Rename DirectBackend fields

ch1bo committed:
Draft usage of ChainBackend

v0d1ch committed:
Unify blockfrost qurying functions to match cardano ones

Being consistent with the parameters and naming should help out.

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Start using CardanoChainConfig

Still keeping around DirectChainConfig until we decide how to
abstract away the details (typeclass, handle pattern, etc.)

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Add class for abstract operations over ChainBackend

Implement queryScriptRegistry and queryGenesisParameters
to see how it looks in action.

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Add queryNetworkId and loadChainContext using it

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Add more queries to make mkTinyWallet work with CardanoChainConfig

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Add queryTimeHandle to BackendOps and start withDirectChain port

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Use generic-lens to disambiguate fields in OptionsSpec

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Add blockfrost backend for chain following

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Connect CardanoChainConfig in Options

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Remove Direct option from hydra-node

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Compile hydra-cluster

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Properly convert to cardano value

We were not handling correctly PolicyId/AssetName conversions and
we were also disregarding any datums/scripts.

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Draft a test for blockfrost integration

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Remap needed test functions to work with blockfrost

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Improve Blocfrost.Client to run in BlockfrostClientT

Add await and run all blockfrost functions in one do block to reduce
posibility of things going wrong but still no luck.

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Fix waiting logic for blockfrost

Added function to wait for specific UTxO. I'd like to refactor
blockfrost client and keep only the least amount of code possible.

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Reverse steps and memory in pparams

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Use CostModelsRaw to construct plutus cost models

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Add code to actually post txs from Direct chain blockfrost component

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Faucet needs to wait for _user_ utxo too

Also a small refactor in the faucet.

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Add timeout to queryUTxO/queryUTxOByTxIn

Make queryUTxOByTxIn take in tx hash as Text
Increase utxo amounts in test to cover for script execution.
Small refactors here and there

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Fix the blockfrost observation

Problem was that we were starting from scratch each time because of
retry. Using TMVar to record block hash solved this problem.

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Add proper metadata to abort tx

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Improve the typeclass design

This slightly improves the design by adding BackendOps
instances for underlying types and then have one instance for
ChainBackend. Too bad we can't derive this instance (yet).

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Make all packages compile again

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Expose submitTransaction for blockfrost

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Alter changelog, remove deleted logs.yaml...

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Constrain blockfrost-client dependency to hydra-node

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Remove generic-lens dependency from hydra-cluster

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Rebase changes

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

ch1bo commented:
@v0d1ch Changed target of this PR to `master` as it seems to include the #1934 changes now (instead of building on top of them; that other branch was not up-to-date with `master` when I reviewed)

ch1bo commented on CHANGELOG.md:
This seems to be messed up now (in commit c36c59de98bc7379bf68f1d93c4c558031eea675)

ch1bo commented on hydra-node/src/Hydra/Chain/Blockfrost/Client.hs:
```diff
@@ -168,8 +169,8 @@ unwrapAddress = \case
-toCardanoNetworkMagic = \case
+toCardanoNetworkId :: Integer -> NetworkId
+toCardanoNetworkId = \case
   0 -> Mainnet
```
Where is this integer coming from? network magic of Mainnet is `764824073`. So `toCardanoNetworkId 764824073` should be `Mainnet` if this integer is the network magic

ch1bo commented on hydra-node/src/Hydra/Chain/Blockfrost/Client.hs:
```diff
@@ -321,29 +322,85 @@ toCardanoGenesisParameters bfGenesis =
+mkEraHistory :: BlockfrostClientT IO EraHistory
+mkEraHistory = do
+  eras' <- Blockfrost.getNetworkEras
+  let eras = filter withoutEmptyEra eras'
```
```suggestion
  let eras = filter isEmptyEra eras'
```

ch1bo commented on hydra-node/src/Hydra/Chain/Blockfrost/Client.hs:
```diff
@@ -321,29 +322,85 @@ toCardanoGenesisParameters bfGenesis =
+  let summary = mkEra <$> eras
+  case nonEmptyFromList summary of
+    Nothing ->
+      liftIO $ throwIO $ BlockfrostError "Failed to create EraHistory."
```
Should: Not use `Text` in errors, but try to enumerate the things that can happen. Here we could say `throwIO EmptyEraHistory` or so

ch1bo commented on hydra-node/src/Hydra/Chain/Blockfrost/Client.hs:
```diff
@@ -321,29 +322,85 @@ toCardanoGenesisParameters bfGenesis =
+----------------
+
+-- | Query the Blockfrost API for address UTxO and convert to cardano 'UTxO'.
+queryUTxO :: SigningKey PaymentKey -> NetworkId -> BlockfrostClientT IO UTxO
```
Should: not require a signing key, but at least a verification key or even better an address. Then, we would not even need `NetworkId`

```suggestion
queryUTxO :: Address ShelleyAddr -> BlockfrostClientT IO UTxO
```

ch1bo commented on hydra-node/src/Hydra/Chain/Blockfrost/Client.hs:
```diff
@@ -321,29 +322,85 @@ toCardanoGenesisParameters bfGenesis =
+queryGenesis = Blockfrost.getLedgerGenesis
+
+-- | Query the Blockfrost API for 'Genesis' and convert to cardano 'ChainPoint'.
+queryTip :: QueryPoint -> BlockfrostClientT IO ChainPoint
```
Must: need a point to query a point

This function does not make sense (from its explanation and signature). If this is really a `queryTip`, it should have signature

```suggestion
queryTip :: BlockfrostClientT IO ChainPoint
```

ch1bo reviewed (commented):
Reviewed this again before looking at the foll integration #1952. I won't renounce my approval (as these things might change in the other PR), but please consider my comments.

v0d1ch committed:
PR review

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Rebase

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
CI secret management

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Bump wait time for blockfrost query

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
No need to update networkId - this was causing failed tests

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Abstract common information into Hydra.Chain.Cardano

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Use reasonably sized RunOptions generator

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

noonio committed:
Blockfrost chain integration (#1952)

- Final step for the blockfrost integration.

---

<!-- Consider each and tick it off one way or the other -->
* [x] CHANGELOG updated or not needed
* [x] Documentation updated or not needed
* [x] Haddocks updated or not needed
* [x] No new TODOs introduced or explained herafter
